### ISSUE TITLE
Better way to calculate per gene Fst

### ISSUE THREAD

**Issue by madzayasodara** [ANGSD/angsd#239]
Hi I'm calculating per gene Fst for all genes.
Which would be the correct and most efficient way of doing this?
I have the Chr startpos endpos for all genes.
Do I have to calculate saf.idx for each gene using the -sites argument, and then go from there to calculate each genes' .ml , fst.idx, and finally do the stats?
Or I can use the whole Chr saf.idx and whole Chr .ml and whole Chr fst.idx, I have already calculated to just run the stats using the -sites?

**Comment by nspope** [comment]
Of those two options I would use the genome-wide SFS/SAF (explanation below) -- then get numerator and denominator of estimator per site by `realSFS fst print` as documented [here](http://www.popgen.dk/angsd/index.php/Fst). You would then sum numerator and denominator across sites in a region (gene), and take the ratio of the sums to get the estimate. Further I would recommend the [Bhatia et al (2013)](https://www.ncbi.nlm.nih.gov/pubmed/23861382) Fst estimator, as it seems to be more stable -- less influenced by mismatched sample size across demes -- than the default Weir & Cockerham estimator. `realSFS` will output the Bhatia estimator by adding the flag `-whichFst 1`.

**Why ...** Fst is a function of allele frequency. The SAF file contains the likelihood of the sequence data at a given site, across all possible sample allele frequencies. Assume we have prior knowledge about allele frequencies (e.g. we know the true site frequency spectrum). `realSFS` calculates Fst at a site as the posterior expectation given the likelihood, prior, and Bayes' rule. In practice, the prior SFS is not known but is instead estimated from the data. The idea is that with a large number of sites, this estimated "prior" will be reasonably accurate. The net effect is to share information across sites so that the per-site/per-region Fst estimates have less error, on average. But because the "prior" is estimated from the data, it is noisy. With a very large number of sites spread across SFS bins, the noise should be negligible. However with only a short sequence (e.g. sequences per gene where only a few coding sequences are variable) the error in the SFS could be quite dramatic, and this would be propagated into your Fst estimates. 

In summary, estimating the prior "globally" may introduce some bias per gene -- by effectively assuming that the SFS (slash evolutionary process) is stationary across the genome -- but if the per gene sequences are relatively short (and/or there are a lot of samples), then on average I think the per-gene Fst estimates should be more accurate.

**Comment by madzayasodara** [comment]
Got it! Just to clarify... You mean I should do all the steps bellow using the whole genome 
#this is with 2pops
#first calculate per pop saf for each populatoin
../angsd -b list1  -anc hg19ancNoChr.fa -out pop1 -dosaf 1 -gl 1
../angsd -b list2  -anc hg19ancNoChr.fa -out pop2 -dosaf 1 -gl 1
#calculate the 2dsfs prior
../misc/realSFS pop1.saf.idx pop2.saf.idx >pop1.pop2.ml
#prepare the fst for easy window analysis etc
../misc/realSFS fst index pop1.saf.idx pop2.saf.idx -sfs pop1.pop2.ml -fstout here -whichFst 1
#print the per site FST

Which will give
                               
NC_007897.1     85      0.000000        0.000001
NC_007897.1     86      0.000000        0.000001
NC_007897.1     87      0.000000        0.000001
NC_007897.1     88      0.000000        0.000001
NC_007897.1     89      0.000000        0.000001
NC_007897.1     90      0.000000        0.000001
NC_007897.1     91      0.000000        0.000001
NC_007897.1     92      0.000000        0.000001
NC_007897.1     93      0.000000        0.000001
NC_007897.1     94      0.000000        0.000001
Where cols are chr    pos   A     B

and 
then calc the weighted fst for the genes regions by doing the ratio between the sum of As and the sum of B. Where A is the alpha from the reynolds 1983 (or Bhatia) and B is the alpha + beta.
 Thanks!

**Comment by nspope** [comment]
Correct. The net result is that the per-gene Fst will be shrunk towards the genome-wide Fst, which will (hopefully) reduce variance in the estimates.

**Comment by madzayasodara** [comment]
Awesome! 
Would the same reasoning apply to calc Tajimas'D and other stuff as in:
do the following step for the using the saf from whole genome
./misc/realSFS out.saf.idx -P 24 > out.sfs
./angsd -bam bam.filelist -out out -doThetas 1 -doSaf 1 -pest out.sfs -anc chimpHg19.fa -GL 1
Then do stats using -sites for each gene 
./misc/thetaStat do_stat out.thetas.idx -sites chr startpos endpos
or should I actually calc thethas for each gene individually 
./angsd -bam bam.filelist -out out -doThetas 1 -doSaf 1 -pest out.sfs -anc chimpHg19.fa -GL 1 -sites chr startpos endpos
and do the stats for each genes's thetas.idx 
./misc/thetaStat do_stat out.thetas.idx

Thanks!

**Comment by nspope** [comment]
You need to sum the thetas within a gene and then calculate Tajima's D, etc. This is because the theta estimators are linear functions of the SFS and so are additive: the sum of a theta estimator from two distinct regions is the estimator for the combined region. Tajima's D (& other similar neutrality stats) is essentially the difference between two theta estimators, but is standardized by the standard deviation of the difference, and so is not additive.

I don't think `thetaStat do_stat` takes a `-sites` argument -- I believe (?) you can only specify chromosome and window size, etc. (but I am unable to check at the moment). While you can probably finagle something on a per-gene basis using `angsd -sites`, it is probably just easier to use `thetaStat print` on the whole shebang to get the log'd theta estimators at each site. Then you can manually exponentiate, sum within desired subsets of sites, and calculate D. 

For example, say you do `thetaStat print my.thetas.idx -r chr1 >logThetas.chr1`, and your coordinates for those genes in "chr1" are in a bed file (e.g. start coordinate is 0-based, end is 1-based)
```
theta = read.table("logThetas.chr1", header=T, comment.char="")
genes = read.table("genes.chr1.bed")[,1:3]
N     = ?? #number of haploid sequences

#coef needed for calculation of D
a1 = sum(1/1:(N-1));  a2 = sum((1/1:(N-1))^2)
b1 = (N+1)/(3*(N-1)); b2 = 2*(N^2 + N + 3)/(9*N*(N-1))
c1 = b1 - 1/a1;       c2 = b2 - (N+2)/(a1*N) + a2/a1^2
e1 = c1/a1;           e2 = c2/(a1^2+a2)

#loop over genes and calculate thetas, number variant sites, D
genes$thetaW = NA
genes$pi = NA
genes$S = NA
genes$TajimaD = NA
for (i in 1:nrow(genes))
{
  sites = (theta$Pos > genes[i,2] & theta$Pos <= genes[i,3])
  genes$thetaW[i]  = sum(exp(theta$Watterson[sites]))
  genes$pi[i]      = sum(exp(theta$Pairwise[sites]))
  genes$S[i]       = a1*genes$thetaW[i]
  genes$TajimaD[i] = (genes$pi[i] - genes$thetaW[i]) / sqrt(e1*genes$S[i] + e2*genes$S[i]^2)
}
head(genes)
```

or something similar

**Comment by madzayasodara** [comment]
Got it! Thanks!
